src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -856,6 +858,100 @@ public ImmutableArray<T> InsertRange(int index, ReadOnlySpan<T> items)
             return self.InsertSpanRangeInternal(index, items);
         }
 
+        /// <summary>
+        /// Removes the specified values from this list.
+        /// </summary>
+        /// <param name="items">The items to remove if matches are found in this list.</param>
+        /// <param name="equalityComparer">
+        /// The equality comparer to use in the search.
+        /// </param>
+        /// <returns>
+        /// A new list with the elements removed.
+        /// </returns>
+        public ImmutableArray<T> RemoveRange(ReadOnlySpan<T> items, IEqualityComparer<T>? equalityComparer = null)
+        {
+            var self = this;
+            self.ThrowNullRefIfNotInitialized();
+
+            if (items.IsEmpty || self.IsEmpty)
+            {
+                return self;
+            }
+
+            if (items.Length == 1)
+            {
+                return self.Remove(items[0], equalityComparer);
+            }
+
+#nullable disable
+            var itemsMultiSet = new Dictionary<T, int>(equalityComparer);
+#nullable restore
+            int nullValueCount = 0;
+            foreach (ref readonly T item in items)
+            {
+                if (item == null)
+                {
+                    nullValueCount++;
+                }
+                else
+                {
+#if NET6_0_OR_GREATER
```

Comment:
Nit: it might be worth exploring if the ifdef'ed logic can be extracted into a helper class, roughly:
```C#
private struct MultiSet<T>
{
     private int _nulls;
     private Dictionary<T, int>? _store;

     public void Add(T value)
     {
#if NET6_0_OR_GREATER
           // insert multiset encoding here
#else
#endif
     }

     public bool TryRemove(T value)
     {
#if NET6_0_OR_GREATER
           // insert multiset encoding here
#else
#endif
     }
}
```
which would allow for cleaner expression of the `RemoveRange` implementation:
```C#
var multiset = new MultiSet<T>(equalityComparer);
foreach (T item in items)
{
      multiset.Add(item);
}

List<int>? indicesToRemove = null;
T[] selfArray = self.array!;
for (int i = 0; i < selfArray.Length; i++)
{
     if (multiSet.TryRemove(selfArray[i]))
     {
          (indicesToRemove ??= new()).Add(i);
     }
}
```

Reply:
This would also remove duplication when applying the same algorithm in the `IEnumerable<T>` overload.

Reply:
If it is `IEnumerable<T>`, can we dare to move it to that `Dictionary`? Because we cannot make sure its size. What if the actual count is very big ?
If yes, maybe need to change the implementation like [mentioned ](https://github.com/dotnet/runtime/pull/61196#discussion_r789574287)above ?

Reply:
That's a good point. It would be a pathological use case regardless, but if we want to bring the same logic here we could type test for `ICollection.Count` (or use Enumerable.TryGetNonEnumeratedCount) and fall back to the existing implementation for arbitrary (or large enough) enumerables.

Reply:
Got it. Besides of this, based on [test result](https://github.com/dotnet/runtime/pull/61196#discussion_r793289310), we also need to care: 
select (dictionary) fast path when current `equalityComparer `is a default one, otherwise fallback to existing implementation also ?

Reply:
Why would we need the slow path when a non-default equality comparer has been supplied?